pilot/pkg/model/authentication.go

```diff
@@ -247,7 +247,7 @@ func getConfigsForWorkload(rootNamespace string, configsByNamespace map[string][
 				switch cfg.GroupVersionKind {
 				case gvk.RequestAuthentication:
 					ra := cfg.Spec.(*v1beta1.RequestAuthentication)
-					should := selectionOpts.ShouldAttachPolicy(cfg.GroupVersionKind, cfg.NamespacedName(), ra)
+					should := selectionOpts.WithRootNamespace(rootNamespace).ShouldAttachPolicy(cfg.GroupVersionKind, cfg.NamespacedName(), ra)
```

Comment:
I am a bit concerned by this `WithRootNamespace` here. Why are mutating the matcher here in a loop? It will then imapct everything outside this function, too. We should probably set it when we actually build the selection opts  -- it looks like we have access to meshConfig there so should be fine.

The changes to mutate are generally problematic too. the code currently assumes non-pointer receivers which do a copy. So when we do something like `WithService` we are making a copy with that services; now it seems we will be mutating it permanently which, if not already a bug, is an accident waiting to happen

Reply:
If you are going to do it we need to make sure we only do it at construction time, and not at each usage.

otherwise we have something like...

```
opts := setupOpts
in some func {  opts.WithService(); use opts }
in some other func { use opts // OOPS, we accidentally have the last used service still }
```

Reply:
All the pass by reference was in a single commit and I dropped that commit. If you think passing by reference should still be done and I just did it incorrectly I can rework it though

Reply:
Good catch; my thought was that pass by reference was a bit more ergonomic, but if it requires a larger refactor, it's probably not worth it